### ISSUE TITLE
Source to target mesh + texture

### ISSUE THREAD

**Issue by cianeastwood** [facebookresearch/pytorch3d#22]
Deforming source to target texture along with the mesh – if possible! Textured interpolation between two objects would be extremely useful for evaluating inverse-graphics models.

**Comment by nikhilaravi** [comment]
@cianeastwood to clarify do you mean source to target mesh deformation with texture using differentiable rendering or with 3D loss functions? The demo in the tutorials is using the 3D mesh losses for the deformation. Do you want to add texture to this example?

**Comment by cianeastwood** [comment]
@nikhilaravi yes I mean with 3D loss functions, as in the example you have from a sphere to a dolphin. Yes it would be great adding texture to this example, allowing the object to be rendered, at various points of the deformation, with a sensible texture.

**Comment by DemonElite** [comment]
is the Texture can be differentiable?  like "Example 3: Optimizing textures" in the [neural_renderer](https://github.com/hiroharu-kato/neural_renderer)  project

**Comment by nikhilaravi** [comment]
@DemonElite yes the texture is differentiable i.e. if you create a tensor for e.g. vertex_rgb colors which has `requires_grad=True` then you should be able to propagate gradients. 

We are currently working on adding support for a per-face texture similar to Neural Mesh Renderer.

**Comment by gattia** [comment]
I have a similar question, I think? 

I have meshes (typically work on them in vtk) that have scalar values associated with each vertex in the mesh. These values are information about the thickness of cartilage on the mesh (a bone). I am looking to use the thickness (scalar) values in a loss function so that a registration (like the dolphin example) takes these into account. 

I cant fully tell from the tex documentation if this is possible. Any thoughts or ideas of where to look would be greatly appreciated. 

Thanks! 

Anthony.

**Comment by nikhilaravi** [comment]
@gattia we are updating the texturing API such that instead of storing a color per vertex as a texture you can store any D dimensional feature.  This might help with your use case. We are currently working on adding support for this.

**Comment by gattia** [comment]
@nikhilaravi Thanks for the update. That would be awesome. 

Just so you have an idea of what I am thinking, I assume that I can create some sort of loss that determines if the "scalar values" or in this case a single texture value between close points on mesh 1 matches mesh 2. 

pytorch3d seems like a potentially great solution for medical image/mesh analyses (registration to start), but the problem I saw originally with the sphere to dolphin was that its good at making any object appear like another object. So, it's possible that it might move points around and have good looking alignment but one anatomical vertex from mesh 1 might not align with the right anatomical vertex of mesh 2. So, I wondered/thought if assigning values to the vertices that coincide with anatomy could help constrain the registration to ensure that for example the inside of the knee stays as the inside of the knee even after registration. 

If this isn't clear but is of use I'm happy to expand more. 

The repo is excellent though, and I've already done playing around with methods like those described in the sphere to dolphin as a method of registering meshes together or of just smoothing a mesh surface.

Kudos!

**Comment by cianeastwood** [comment]
@nikhilaravi – is there any update on this? Completely understand that it's probably not very high on the priority list right now... just curious as I have a paper waiting in the wings! 😃

**Comment by nikhilaravi** [comment]
@cianeastwood we have added an example tutorial for how to deform and fit a sphere mesh to a target mesh along with texture: https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/fit_textured_mesh.ipynb. Hopefully this helps with your project!

@DemonElite, @gattia, the textures API was refactored in https://github.com/facebookresearch/pytorch3d/commit/a3932960b31ff07e942a54e4608eae6ba12bf40a. So now you can store textures as a per face texture map (`TexturesAtlas`), or a D dimensional feature per vertex (`TexturesVertex`).

**Comment by nikhilaravi** [comment]
Please close this issue if you have no further questions!
